bfvmm/src/hve/arch/intel_x64/ept/memory_map.cpp

```diff
@@ -0,0 +1,372 @@
+            "is not mapped at the 2MB level");
+    }
+    if (epte::is_leaf_entry(pde)) {
+        return pde::page_address::get_effective_address(pde, gpa);
+    }
+
+    auto &pte = this->gpa_to_pte(gpa, pde);
+    if (!epte::is_present(pte)) {
+        throw std::runtime_error("gpa_to_hpa: failed to resolve gpa->epte, gpa "
+            "is not mapped at the 4KB level");
+    }
+    if (epte::is_leaf_entry(pte)) {
+        return pte::page_address::get_effective_address(pte, gpa);
+    }
+
+    throw std::runtime_error("gpa_to_hpa: extended page tables corrupted");
+}
+
+std::vector<memory_descriptor>
+memory_map::to_mdl() const
+{
+    std::vector<memory_descriptor> mdl{0};
+    mdl.push_back({m_pml4_hpa, m_pml4_hva, MEMORY_TYPE_R | MEMORY_TYPE_W});
+
+    mdl = this->to_mdl(mdl, reinterpret_cast<epte_t *>(m_pml4_hva));
+    return mdl;
+}
+
+hpa_t
+memory_map::hpa() const
+{
+    return m_pml4_hpa;
+}
+
+hpa_t
+memory_map::allocate_page_table()
+{
+    auto pt_hva = new epte_t [page_table::num_entries]();
+    auto pt_hpa = g_mm->virtptr_to_physint(pt_hva);
+    return pt_hpa;
+}
+
+void
+memory_map::allocate_page_table(epte_t &entry)
+{
+    auto pt_hpa = this->allocate_page_table();
+
+    epte::read_access::enable(entry);
+    epte::write_access::enable(entry);
+    epte::execute_access::enable(entry);
+    epte::memory_type::set(entry, epte::memory_type::wb);
+    epte::set_hpa(entry, pt_hpa);
+}
+
+void
+memory_map::free_page_table(epte_t &entry)
+{
+    auto pt_hpa = epte::hpa(entry);
+    auto pt_hva = g_mm->physint_to_virtptr(pt_hpa);
+    auto page_table = static_cast<epte_t *>(pt_hva);
+
+    auto pt_view = gsl::make_span(page_table, page_table::num_entries);
+    for (auto pte : pt_view) {
+        if (epte::is_present(pte) && !epte::is_leaf_entry(pte)) {
+            this->free_page_table(pte);
+        }
+    }
+
+    epte::clear(entry);
+    delete[] page_table;
+}
+
+void
+memory_map::map_entry_to_page_frame(epte_t &entry, hpa_t hpa)
+{
+    epte::read_access::enable(entry);
+    epte::write_access::enable(entry);
+    epte::memory_type::set(entry, epte::memory_type::wb);
+    epte::entry_type::enable(entry);
+    epte::set_hpa(entry, hpa);
+}
+
+epte_t &
+memory_map::gpa_to_pml4e(gpa_t gpa)
+{
+    auto pml4e_offset = gpa::pml4_index::get_offset(gpa);
+    auto pml4_hpa = m_pml4_hpa;
+    auto pml4_hva = g_mm->physint_to_virtint(pml4_hpa);
+    auto pml4e_hva = pml4_hva + pml4e_offset;
+    return *reinterpret_cast<epte_t *>(pml4e_hva);
+}
+
+epte_t &
+memory_map::gpa_to_pdpte(gpa_t gpa, epte_t &pml4e)
+{
+    auto pdpte_offset = gpa::pdpt_index::get_offset(gpa);
+    auto pdpt_hpa = epte::hpa(pml4e);
+    auto pdpt_hva = g_mm->physint_to_virtint(pdpt_hpa);
+    auto pdpte_hva = pdpt_hva + pdpte_offset;
+    return *reinterpret_cast<epte_t *>(pdpte_hva);
+}
+
+epte_t &
+memory_map::gpa_to_pde(gpa_t gpa, epte_t &pdpte)
+{
+    auto pde_offset = gpa::pd_index::get_offset(gpa);
+    auto pd_hpa = epte::hpa(pdpte);
```

Comment:
same thing here, to simplify, just toss the return of epte::hpa(pdpte) into the input of the next line since you have enough horizontal space

Reply:
Don't think I'll change these, it is more readable to me to see each step explicitly separated

Reply:
I agree with Jared on this